Update and comment demo (#39)
adrhill/iclr2025-sparse-autodiff: _posts/2025-04-28-sparse-autodiff.md

### Preparation

> Sparsity pattern detection and matrix coloring are performed in a so-called "preparation step", whose output can be r**eused across several calls** to `jacobian` (as long as the pattern stays the same).

Reviewer: ```suggestion
Sparsity pattern detection and matrix coloring are performed in a so-called "preparation step", whose output can be **reused across several calls** to `jacobian` (as long as the pattern stays the same).
```

4
```

### Coloring visualization

We just saw that there is a discrepancy between the number of different colors $c$ and the input size $n$.

> This discrepancy typically gets larger as the input grows.

Reviewer: ```suggestion
This discrepancy typically gets larger as the input dimension grows.
```

The larger the ratio $n / c$, the more useful ASD will be.

{% include figure.html path="assets/img/2025-04-28-sparse-autodiff/demo/banded.png" class="img-fluid" %}
<div class="caption">
    Figure ??: Coloring numbers are often agnostic to the input size.  
</div>

On Figure ??, we see the Jacobians of `iter_diff` for several values of $n$ and $k$.
The main takeaway is that the number of colors does not depend on the dimension $n$, only on the number of iterations $k$.
In fact, `iter_diff` with $k$ iterations gives rise to a banded Jacobian with $k+1$ bands, for which we know that the optimal coloring uses as many colors as bands.

> For this particular case, the greedy coloring finds the optimal solution.

Reviewer: First text, then figure, no?

### Performance benefits

Here we present a benchmark for the Jacobian of `iter_diff` with varying $n$ and fixed $k$.

> Our goal is to find out when ASD becomes more advantageous than standard AD.

Reviewer: ```suggestion
Our goal is to find out when ASD becomes more performant than standard AD.
```

### Performance benefits

Here we present a benchmark for the Jacobian of `iter_diff` with varying $n$ and fixed $k$.
Our goal is to find out when sparse differentiation becomes relevant.
Benchmark data can be generated with the following code:

```julia
using DifferentiationInterfaceTest
scenarios = [
    Scenario{:jacobian, :out}(iter_diff, rand(n); contexts=(Constant(k),))
    for n in round.(Int, 10 .^ (0:0.3:4))
]
data = benchmark_differentiation(
    [dense_backend, sparse_backend],
    scenarios;
    benchmark=:full
)
```

It gives rise to the following performance curves (lower is better):

{% include figure.html path="assets/img/2025-04-28-sparse-autodiff/demo/benchmark.png" class="img-fluid" %}
<div class="caption">
    Figure 22: Performance benefits of sparsity  
</div>

As we can see on Figure 22, there are three main regimes:

1. For very small inputs, we gain nothing by leveraging sparsity.
2. For medium-sized inputs, sparsity handling is only useful if we can amortize the cost of detection and coloring.
3. For very large inputs, even the overhead of detection and coloring is worth paying as part of a sparse Jacobian computation.

> Importantly, sparsity can yield an **asymptotic speedup** and not just a constant one.

Reviewer: Can we find a better term than "asymptotic speedup"? I have trouble wrapping my head around it, since it assumes some "reference performance" that the speed-up is measured w.r.t.

Author: yes, the reference performance is that of standard AD
